### ISSUE TITLE
Failed assertion when receiving key repeat event with no keyboard focus

### ISSUE THREAD

**Issue by jamii** [glfw/glfw#2727]
(This is against https://github.com/tiawl/glfw.zig/commit/a611b31721c05010e7cbfcaea28e8075ff9ed40c, which has a different build setup but should otherwise be vanilla glfw 3.4. That might make repro hard, but I thought it was worth recording this somewhere anyway.)

I'm seeing a few crashes per day in:

```
focus: /home/jamie/focus/deps/glfw/glfw/src/input.c:274: void _glfwInputKey(_GLFWwindow *, int, int, int, int): Assertion `window != NULL' failed.
Set Windows Application User Model ID (AUMID) { AUMID: 'org.whispersystems.signal-desktop' }
NODE_ENV production
NODE_CONFIG_DIR /nix/store/s81a57m4lnr0l32n29v2ax22vqypw2vh-signal-desktop-7.56.1/share/signal-desktop/app.asar/config
NODE_CONFIG {}
ALLOW_CONFIG_MUTATIONS undefined
HOSTNAME vessel
NODE_APP_INSTANCE undefined
SUPPRESS_NO_CONFIG_WARNING undefined
SIGNAL_ENABLE_HTTP undefined
userData: /home/jamie/.config/Signal
config/get: Successfully read user config file
config/get: Successfully read ephemeral config fil
```

From the coredump, it looks like the problem is that glfw sometimes receives a key repeat event while there is no keyboard focus.

```
>>> bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007ffff7c99d43 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:78
#2  0x00007ffff7c413c6 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff7c2893a in __GI_abort () at abort.c:79
#4  0x00007ffff7c2885e in __assert_fail_base (fmt=0x7ffff7dbdf20 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x103c61b "window != NULL",
    file=file@entry=0x10385f5 "/home/jamie/focus/deps/glfw/glfw/src/input.c", line=line@entry=274, function=function@entry=0x1057d3a "void _glfwInputKey(_GLFWwindow *, int, int, int, int)")
    at assert.c:96
#5  0x00007ffff7c395a6 in __assert_fail (assertion=0x103c61b "window != NULL", file=0x10385f5 "/home/jamie/focus/deps/glfw/glfw/src/input.c", line=274,
    function=0x1057d3a "void _glfwInputKey(_GLFWwindow *, int, int, int, int)") at assert.c:105
#6  0x0000000001230864 in _glfwInputKey (window=0x0, key=257, scancode=28, action=1, mods=0) at /home/jamie/focus/deps/glfw/glfw/src/input.c:274
#7  0x0000000001297ab8 in handleEvents (timeout=0x7fffffffb3a0) at /home/jamie/focus/deps/glfw/glfw/src/wl_window.c:1272
#8  0x0000000001296116 in _glfwPollEventsWayland () at /home/jamie/focus/deps/glfw/glfw/src/wl_window.c:2641
#9  0x000000000124f207 in glfwPollEvents () at /home/jamie/focus/deps/glfw/glfw/src/window.c:1142
#10 0x0000000001121a45 in lib.focus.App.frame (self=0x1b6f950) at /home/jamie/focus/lib/focus.zig:349
#11 lib.focus.run (server_socket=...) at /home/jamie/focus/lib/focus.zig:182
#12 0x000000000111ed1d in focus.main () at focus.zig:48
#13 start.callMain () at /nix/store/j3hylam3gxzrmzs52jcx18nfljsv1wgb-zig/lib/std/start.zig:647
#14 start.callMainWithArgs () at /nix/store/j3hylam3gxzrmzs52jcx18nfljsv1wgb-zig/lib/std/start.zig:616
#15 start.main (c_argc=<optimized out>, c_argv=<optimized out>, c_envp=<optimized out>) at /nix/store/j3hylam3gxzrmzs52jcx18nfljsv1wgb-zig/lib/std/start.zig:631

>>> f 7
#7  0x0000000001297ab8 in handleEvents (timeout=0x7fffffffb3a0) at /home/jamie/focus/deps/glfw/glfw/src/wl_window.c:1272
1272	                    _glfwInputKey(_glfw.wl.keyboardFocus,

>>> p _glfw.wl.keyboardFocus
$2 = (_GLFWwindow *) 0x0
```

This is running against [niri](https://github.com/YaLTeR/niri) 25.05.1 on nixos 25.05.804113.6c64dabd3aa8.

Exe and coredump, if that's helpful:
https://user.fm/files/v2-d73a56a78dfd17e74d59cbf96dde8574/dump
https://user.fm/files/v2-79ec0bda43172cb41d3f1e33f18fd6e7/focus

**Comment by GamesTrap** [comment]
```diff
diff --git a/src/wl_window.c b/src/wl_window.c
index 72c1a402..6457f31e 100644
--- a/src/wl_window.c
+++ b/src/wl_window.c
@@ -1260,31 +1260,35 @@ static void handleEvents(double* timeout)
         }
         else
             wl_display_cancel_read(_glfw.wl.display);
 
         if (fds[KEYREPEAT_FD].revents & POLLIN)
         {
             uint64_t repeats;
 
             if (read(_glfw.wl.keyRepeatTimerfd, &repeats, sizeof(repeats)) == 8)
             {
-                for (uint64_t i = 0; i < repeats; i++)
+                if(_glfw.wl.keyboardFocus)
                 {
-                    _glfwInputKey(_glfw.wl.keyboardFocus,
-                                  translateKey(_glfw.wl.keyRepeatScancode),
-                                  _glfw.wl.keyRepeatScancode,
-                                  GLFW_PRESS,
-                                  _glfw.wl.xkb.modifiers);
-                    inputText(_glfw.wl.keyboardFocus, _glfw.wl.keyRepeatScancode);
+                    for (uint64_t i = 0; i < repeats; i++)
+                    {
+                        _glfwInputKey(_glfw.wl.keyboardFocus,
+                                      translateKey(_glfw.wl.keyRepeatScancode),
+                                      _glfw.wl.keyRepeatScancode,
+                                      GLFW_PRESS,
+                                      _glfw.wl.xkb.modifiers);
+                        inputText(_glfw.wl.keyboardFocus, _glfw.wl.keyRepeatScancode);
+                    }
+
+                    event = GLFW_TRUE;
                 }
 
-                event = GLFW_TRUE;
             }
         }
 
         if (fds[CURSOR_FD].revents & POLLIN)
         {
             uint64_t repeats;
 
             if (read(_glfw.wl.cursorTimerfd, &repeats, sizeof(repeats)) == 8)
                 incrementCursorImage(_glfw.wl.pointerFocus);
         }

```
This may fix the crash, untested.

You may need to set `event = GLFW_TRUE;` even when `_glfw.wl.keyboardFocus` is `NULL`.

**Comment by dougbinks** [comment]
Do you have a multithreaded environment with event polling via `glfwPollEvents()` on a different thread to main and the `glfwInit()` call? If so this is unsupported as per documentation as both must be on the main thread. I'm assuming this isn't the case but I would like to check.

If these are on the main thread then your fix should work but I'd like to double check what's going on.

Is this repeatable enough (pun not intended) to get the assert in the `events.c` test code?

**Comment by jamii** [comment]
I don't have any threads in this program.

I'm not sure how to actually trigger the crash. I just every few hours notice the window is gone and then I find the above crash in the logs. I tried just now mashing buttons while switching workspaces but no luck.

**Comment by dougbinks** [comment]
> I don't have any threads in this program.

Thanks for confirming that.

> I'm not sure how to actually trigger the crash. I just every few hours notice the window is gone and then I find the above crash in the logs. I tried just now mashing buttons while switching workspaces but no luck.

Thanks for trying.

**Comment by dougbinks** [comment]
@GamesTrap Do you want to submit your PR as a potential fix (we'll have to review and test but it looks reasonable at first glance)?

@jamii Are you able to test @GamesTrap's proposed solution?

**Comment by jamii** [comment]
Sure, I have it running at the moment. Will check the logs tomorrow to see if I still get any crashes.

**Comment by GamesTrap** [comment]
> [@GamesTrap](https://github.com/GamesTrap) Do you want to submit your PR as a potential fix (we'll have to review and test but it looks reasonable at first glance)?
> 
> [@jamii](https://github.com/jamii) Are you able to test [@GamesTrap](https://github.com/GamesTrap)'s proposed solution?

I've created a PR for this: https://github.com/glfw/glfw/pull/2732

**Comment by jamii** [comment]
I've seen no more crashes in the week since applying this patch.

**Comment by dougbinks** [comment]
Thanks - I'll take another look at the PR tomorrow and [go through the contributing a bug fix checklist](https://github.com/glfw/glfw/blob/master/docs/CONTRIBUTING.md#contributing-a-bug-fix).

**Comment by dougbinks** [comment]
Given your feedback that you've seen no crashes since applying the path, this should now be fixed with https://github.com/glfw/glfw/commit/506c11ba43b901dbcc4d90449f46de67cf000af4

**Comment by dougbinks** [comment]
Hi @jamii I have a further question about this issue. Whilst we resolved the assert/segfault with PR #2732 this did not identify the root cause of the problem.

When a window looses keyboard focus it resets the key repeat timer, so this issue should not occur in that case. However I have found that if a window with keyboard focus is closed the key repeat timer is not reset (see #2741) and I can replicate this issue that way (without the workaround in PR #2732).

Might a window have closed in your app whilst it held keyboard input focus? For example if you're using Dear ImGui with external viewports then docking a window with another can trigger a window destruction and thus potentially trigger this issue.
